New post: Smaller is better: Q8-Chat, an efficient generative AI experience on Xeon (#1084)
huggingface/blog: generative-ai-models-on-intel-cpu.md

In a nutshell, quantization rescales model parameters to smaller value ranges. When successful, it shrinks your model by at least 2x, without any impact on model accuracy.

You can apply quantization during training, a.k.a quantization-aware training ([QAT](https://arxiv.org/abs/1910.06188)), which generally yields the best results. If you’d prefer to quantize an existing model, you can apply post-training quantization ([PTQ](https://www.tensorflow.org/lite/performance/post_training_quantization#:~:text=Post%2Dtraining%20quantization%20is%20a,little%20degradation%20in%20model%20accuracy.)), a much faster technique that requires very little computing power.

Different quantization tools are available. For example, PyTorch has built-in support for [quantization](https://pytorch.org/docs/stable/quantization.html). You can also use the Hugging Face [Optimum Intel](https://huggingface.co/docs/optimum/intel/index) library, which includes developer-friendly APIs for QAT and PTQ.

## Quantizing LLMs

> 

Reviewer: In the end we used the example from OPT, should change Llama-7B to OPT-13B

Recent studies [[1]](https://arxiv.org/abs/2206.01861)[[2]](https://arxiv.org/abs/2211.10438) show that current quantization techniques don’t work well with LLMs. In particular, LLMs exhibit large-magnitude outliers in specific activation channels across all layers and tokens. Here’s an example with the Llama-7B model. You can see that one of the activation channels has much larger values than all others across all tokens. This phenomenon is visible in all the Transformer layers of the model.

> 

Reviewer: Need to add a title to the figure "OPT-13B Input to layer 3.FC1", we produced this figure ourselves if it matters

<kbd>
  <img src="assets/142_q8chat/pic2.png">
</kbd>
<br>*Source: SmoothQuant*

Now, let’s see how SmoothQuant works when applied to popular LLMs.

## Quantizing LLMs with SmoothQuant

Our friends at Intel have quantized several LLMs with SmoothQuant-O3: OPT [2.7B](https://huggingface.co/facebook/opt-2.7b) and [6.7B](https://huggingface.co/facebook/opt-6.7b) [[5]](https://arxiv.org/pdf/2205.01068.pdf), LLaMA [7B](https://huggingface.co/decapoda-research/llama-7b-hf) [[6]](https://ai.facebook.com/blog/large-language-model-llama-meta-ai/), Alpaca [7B](https://huggingface.co/tatsu-lab/alpaca-7b-wdiff) [[7]](https://crfm.stanford.edu/2023/03/13/alpaca.html), Vicuna [7B](https://huggingface.co/lmsys/vicuna-7b-delta-v1.1) [[8]](https://vicuna.lmsys.org/), BloomZ [7.1B](https://huggingface.co/bigscience/bloomz-7b1) [[9]](https://huggingface.co/bigscience/bloomz). They also evaluated the accuracy of the quantized models using the [Language Model Evaluation Harness](https://github.com/EleutherAI/lm-evaluation-harness).

> The table below presents a summary of their findings. The second column shows the ratio of benchmarks that have improved post-quantization. The third column contain the mean average degradation (a negative value indicates that the benchmark has improved). You can find the detailed results at the end of this post.

Reviewer: Add a \*, (* a negativa ...) So it will be clear the * in the table refers to this comment

> The table below presents a summary of their findings. The second column shows the ratio of benchmarks that have improved post-quantization. The third column contain the mean average degradation (a negative value indicates that the benchmark has improved). You can find the detailed results at the end of this post.

Reviewer: We think that the second column might be misunderstood, maybe it is better to talk about how many tasks preserves accuracy degradation of less than 1% or something like that, or add some clarification that it is ok that we are not improving accross all tasks because we are dealing with lossy compression, or remove it all-together.

<kbd>
  <img src="assets/142_q8chat/table0.png">
</kbd>

As you can see, OPT models are great candidates for SmoothQuant quantization. Models are ~2x smaller compared to pretrained 16-bit models. Most of the metrics improve, and those who don’t are only marginally penalized. 

The picture is a little more contrasted for LLaMA 7B and BloomZ 7.1B. Models are compressed by a factor of ~2x, with about half the task seeing metric improvements. Again, the other half is only marginally impacted, with a single task seeing more than 3% relative degradation.

> The obvious benefit of working with smaller models is a significant reduction in inference latency. Here’s a [video](https://drive.google.com/file/d/19PrWwJo-laXYppRzUZyR7s3M1zd_2Yfe/view?usp=share_link) demonstrating real-time text generation with the Vicuna-7b model, on a single socket Intel Sapphire Rapids CPU with 32 cores and a batch size of 1.

Reviewer: We recorded a new video where it is clear where the compute cycle begins, https://drive.google.com/file/d/1h8C2I4xn1c0HdrzfMqBaYECJyqo5kcJL/view?usp=sharing

The picture is a little more contrasted for LLaMA 7B and BloomZ 7.1B. Models are compressed by a factor of ~2x, with about half the task seeing metric improvements. Again, the other half is only marginally impacted, with a single task seeing more than 3% relative degradation.

The obvious benefit of working with smaller models is a significant reduction in inference latency. Here’s a [video](https://drive.google.com/file/d/1h8C2I4xn1c0HdrzfMqBaYECJyqo5kcJL/view?usp=sharing) demonstrating real-time text generation with the Vicuna-7b model on a single socket Intel Sapphire Rapids CPU with 32 cores and a batch size of 1.

In this example, we ask the model: “*What is the role of Hugging Face in democratizing NLP?*”. This sends the following prompt to the model:
"*A chat between a curious user and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the user's questions. USER: What is the role of Hugging Face in democratizing NLP? ASSISTANT:*"

<kbd>
  [<img src="assets/142_q8chat/pic3.png">](assets/142_q8chat/vicuna-7b-int8-hf-role.mov)
</kbd>

> It only takes ~80 milliseconds to generate the first token. The mean generation time for the next tokens is an amazing ~40 milliseconds. This level of performance definitely makes it possible to run LLMs on CPU platforms, giving customers more IT flexibility and better cost-performance than ever before.

Reviewer 2: replace the line:
"It only takes ~80 milliseconds to generate the first token. The mean generation time for the next tokens is an amazing ~40 milliseconds."
with:
"As you can see in the example with 8bit quantization the latency per word is very low which results in fast generation."

> It only takes ~80 milliseconds to generate the first token. The mean generation time for the next tokens is an amazing ~40 milliseconds. This level of performance definitely makes it possible to to run LLMs on CPU platforms, giving customers more IT flexibility and better cost-performance than ever before.

Reviewer 3: ```suggestion
It only takes ~80 milliseconds to generate the first token. The mean generation time for the next tokens is an amazing ~40 milliseconds. This level of performance definitely makes it possible to run LLMs on CPU platforms, giving customers more IT flexibility and better cost-performance than ever before.
```

## Chat experience on Xeon

> Recently, Clement, the CEO of HuggingFace, said: “*More companies would be better served focusing on smaller, specific models that are cheaper to train and run.*” The emergence of relatively smaller models like Alpaca, BloomZ and Vicuna, open a new opportunity for enterprise to lower the cost of fine-tuning and inference in production. As demonstrated above, high-quality quantization brings high-quality chat experiences to Intel CPU platforms, without the need of running mammoth LLMs and complex AI accelerators. 

Reviewer 3: ```suggestion
Clement, the CEO of HuggingFace, recently said: “*More companies would be better served focusing on smaller, specific models that are cheaper to train and run.*” The emergence of relatively smaller models like Alpaca, BloomZ and Vicuna, open a new opportunity for enterprises to lower the cost of fine-tuning and inference in production. As demonstrated above, high-quality quantization brings high-quality chat experiences to Intel CPU platforms, without the need of running mammoth LLMs and complex AI accelerators. 
```

> Together with Intel, we've hosted on Spaces a new exciting demo called [Q8-Chat](https://huggingface.co/spaces/Intel/Q8-Chat) (pronounced "Cute chat"). Q8-Chat offers you a ChatGPT-like chat experience, while only running on a single socket Intel Sapphire Rapids CPU with 32 cores and a batch size of 1.

Reviewer 3: ```suggestion
Together with Intel, we're hosting a new exciting demo in Spaces called [Q8-Chat](https://huggingface.co/spaces/Intel/Q8-Chat) (pronounced "Cute chat"). Q8-Chat offers you a ChatGPT-like chat experience, while only running on a single socket Intel Sapphire Rapids CPU with 32 cores and a batch size of 1.
```

<kbd>
  <img src="assets/142_q8chat/pic4.png">
</kbd>

## Next steps

> We’re currently working on integrating these amazing new quantization techniques into the Hugging Face [Optimum Intel](https://huggingface.co/docs/optimum/intel/index) library. Once we’re done, you’ll be able to replicate these demos with just a few lines of code.

Reviewer: Can you please add
"... into the Hugging Face [Optimum Intel](https://huggingface.co/docs/optimum/intel/index) library through [Intel Neural Compressor](https://github.com/intel/neural-compressor)."

</kbd>
<br>*Source: SmoothQuant*

The best quantization techniques to date quantize activations token-wise, causing either truncated outliers or underflowing low-magnitude activations. Both solutions hurt model quality significantly. Moreover, quantization-aware training requires additional model training, which is not practical in most cases due to lack of compute resources and data.

SmoothQuant [[3]](https://arxiv.org/abs/2211.10438)[[4]](https://github.com/mit-han-lab/smoothquant) is a new quantization technique that solves this problem. It applies a joint mathematical transformation to weights and activations, which reduces the ratio between outlier and non-outlier values for activations at the cost of increasing the ratio for weights. This transformation makes the layers of the Transformer "quantization-friendly" and enables 8-bit quantization without hurting model quality. As a consequence, SmoothQuant produces smaller, faster models that run well on Intel CPU platforms.

<kbd>
  <img src="assets/142_q8chat/pic2.png">
</kbd>
<br>*Source: SmoothQuant*

Now, let’s see how SmoothQuant works when applied to popular LLMs.

## Quantizing LLMs with SmoothQuant

> Our friends at Intel have quantized several LLMs with SmoothQuant-O3: OPT [2.7B](https://huggingface.co/facebook/opt-2.7b) and [6.7B](https://huggingface.co/facebook/opt-6.7b) [[5]](https://arxiv.org/pdf/2205.01068.pdf), LLaMA [7B](https://huggingface.co/decapoda-research/llama-7b-hf) [[6]](https://ai.facebook.com/blog/large-language-model-llama-meta-ai/), Alpaca [7B](https://huggingface.co/tatsu-lab/alpaca-7b-wdiff) [[7]](https://crfm.stanford.edu/2023/03/13/alpaca.html), Vicuna [7B](https://huggingface.co/lmsys/vicuna-7b-delta-v1.1) [[8]](https://vicuna.lmsys.org/), BloomZ [7.1B](https://huggingface.co/bigscience/bloomz-7b1) [[9]](https://huggingface.co/bigscience/bloomz). They also evaluated the accuracy of the quantized models, using [Language Model Evaluation Harness](https://github.com/EleutherAI/lm-evaluation-harness).

Reviewer 2: add after BloomZ:
and MPT-7B-chat [https://www.mosaicml.com/blog/mpt-7b](url)
